In [1]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv("Healthcare Providers.csv")

In [3]:
print("First 10 rows of the dataset:")
df.head(10)

First 10 rows of the dataset:


index  National Provider Identifier  \
0  8774979                    1891106191   
1  3354385                    1346202256   
2  3001884                    1306820956   
3  7594822                    1770523540   
4   746159                    1073627758   
5  3443985                    1346571551   
6  2137027                    1215943535   
7  6169160                    1629160551   
8  5086226                    1518929124   
9  3900718                    1396781134   

  Last Name/Organization Name of the Provider First Name of the Provider  \
0                                 UPADHYAYULA                  SATYASREE   
1                                       JONES                      WENDY   
2                                    DUROCHER                    RICHARD   
3                                     FULLARD                     JASPER   
4                                    PERROTTI                    ANTHONY   
5                                        PUGH                       JOHN   
6                                     BRUMITT                        TOM   
7                                   GALBREATH                     RONALD   
8                                       BOONE                      RALPH   
9                                 METWEST INC                        NaN   

  Middle Initial of the Provider Credentials of the Provider  \
0                            NaN                        M.D.   
1                              P                        M.D.   
2                              W                         DPM   
3                            NaN                          MD   
4                              E                          DO   
5                              R                         DPT   
6                              B                          DO   
7                              G                        M.D.   
8                              M                        D.O.   
9                            NaN                         NaN   

  Gender of the Provider Entity Type of the Provider  \
0                      F                           I   
1                      F                           I   
2                      M                           I   
3                      M                           I   
4                      M                           I   
5                      M                           I   
6                      M                           I   
7                      M                           I   
8                      M                           I   
9                    NaN                           O   

  Street Address 1 of the Provider Street Address 2 of the Provider  ...  \
0                1402 S GRAND BLVD                   FDT 14TH FLOOR  ...   
1                  2950 VILLAGE DR                              NaN  ...   
2                20 WASHINGTON AVE                          STE 212  ...   
3               5746 N BROADWAY ST                              NaN  ...   
4                 875 MILITARY TRL                        SUITE 200  ...   
5                 504 ALBEMARLE SQ                              NaN  ...   
6                  70 DOCTORS PARK                              NaN  ...   
7            12522 E. LAMBERT ROAD                          SUITE D  ...   
8                    1215 DUNN AVE                              NaN  ...   
9                   695 S BROADWAY                              NaN  ...   

  HCPCS Code                                  HCPCS Description  \
0      99223  Initial hospital inpatient care, typically 70 ...   
1      G0202  Screening mammography, bilateral (2-view study...   
2      99348  Established patient home visit, typically 25 m...   
3      81002                            Urinalysis, manual test   
4      96372  Injection beneath the skin or into muscle for ...   
5      97016  Application of blood vessel compression or dec...   
6      20610  Aspiration and/or injection of large join

In [4]:
print("Shape of the Dataset:")
df.shape

Shape of the Dataset:


(100000, 27)

In [5]:
print("\nDataset information")
df.info()


Dataset information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                                                    Non-Null Count   Dtype  
---  ------                                                    --------------   -----  
 0   index                                                     100000 non-null  int64  
 1   National Provider Identifier                              100000 non-null  int64  
 2   Last Name/Organization Name of the Provider               100000 non-null  object 
 3   First Name of the Provider                                95745 non-null   object 
 4   Middle Initial of the Provider                            70669 non-null   object 
 5   Credentials of the Provider                               92791 non-null   object 
 6   Gender of the Provider                                    95746 non-null   object 
 7   Entity Type of the Provider                               100000 non-nul

## These Five Columns Have NULL Values
* 3   First Name of the Provider                                95745 non-null   object 
* 4   Middle Initial of the Provider                            70669 non-null   object 
* 5   Credentials of the Provider                               92791 non-null   object 
* 6   Gender of the Provider                                    95746 non-null   object 
* 9   Street Address 2 of the Provider                          40637 non-null   object

In [7]:
print("\nStatistical summary of the dataset:")
print(df.describe(include='all'))


Statistical summary of the dataset:
               index  National Provider Identifier  \
count   1.000000e+05                  1.000000e+05   
unique           NaN                           NaN   
top              NaN                           NaN   
freq             NaN                           NaN   
mean    4.907646e+06                  1.498227e+09   
std     2.839633e+06                  2.874125e+08   
min     2.090000e+02                  1.003001e+09   
25%     2.458791e+06                  1.245669e+09   
50%     4.901266e+06                  1.497847e+09   
75%     7.349450e+06                  1.740374e+09   
max     9.847440e+06                  1.993000e+09   

       Last Name/Organization Name of the Provider First Name of the Provider  \
count                                       100000                      95745   
unique                                       42820                      13022   
top                                          PATEL                    M

In [8]:
print("\nMissing values and their percentages:")
missing_values = df.isnull().sum()
missing_percentages = (missing_values / len(df)) * 100
missing_data = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage': missing_percentages
})
print(missing_data)


Missing values and their percentages:
                                                    Missing Values  Percentage
index                                                            0       0.000
National Provider Identifier                                     0       0.000
Last Name/Organization Name of the Provider                      0       0.000
First Name of the Provider                                    4255       4.255
Middle Initial of the Provider                               29331      29.331
Credentials of the Provider                                   7209       7.209
Gender of the Provider                                        4254       4.254
Entity Type of the Provider                                      0       0.000
Street Address 1 of the Provider                                 0       0.000
Street Address 2 of the Provider                             59363      59.363
City of the Provider                                             0       0.000
Zip Code of t

In [9]:
cleaned_df = df.dropna(thresh=len(df) * 0.5, axis=1)
print("Shape after removing columns with more than 50% missing values:")
print(cleaned_df.shape)

Shape after removing columns with more than 50% missing values:
(100000, 26)


## Column - Street Address 2 of the Provider have Null Values more than 50% so that column is dropped

In [11]:
numerical_columns = cleaned_df.select_dtypes(include=['int64', 'float64']).columns
print("Numerical Columns:")
print(numerical_columns)

categorical_columns = cleaned_df.select_dtypes(include=['object']).columns
print("\nCategorical Columns:")
print(categorical_columns)

Numerical Columns:
Index(['index', 'National Provider Identifier', 'Zip Code of the Provider'], dtype='object')

Categorical Columns:
Index(['Last Name/Organization Name of the Provider',
       'First Name of the Provider', 'Middle Initial of the Provider',
       'Credentials of the Provider', 'Gender of the Provider',
       'Entity Type of the Provider', 'Street Address 1 of the Provider',
       'City of the Provider', 'State Code of the Provider',
       'Country Code of the Provider', 'Provider Type',
       'Medicare Participation Indicator', 'Place of Service', 'HCPCS Code',
       'HCPCS Description', 'HCPCS Drug Indicator', 'Number of Services',
       'Number of Medicare Beneficiaries',
       'Number of Distinct Medicare Beneficiary/Per Day Services',
       'Average Medicare Allowed Amount', 'Average Submitted Charge Amount',
       'Average Medicare Payment Amount',
       'Average Medicare Standardized Amount'],
      dtype='object')


In [12]:
# List of categorical columns to fill with mode
categorical_columns = [
    'First Name of the Provider',
    'Middle Initial of the Provider',
    'Credentials of the Provider',
    'Gender of the Provider'
]

# Fill missing values for each column
for column in categorical_columns:
    cleaned_df[column].fillna(cleaned_df[column].mode()[0], inplace=True)

In [13]:
print("\nMissing values after imputation:")
print(cleaned_df.isnull().sum())



Missing values after imputation:
index                                                       0
National Provider Identifier                                0
Last Name/Organization Name of the Provider                 0
First Name of the Provider                                  0
Middle Initial of the Provider                              0
Credentials of the Provider                                 0
Gender of the Provider                                      0
Entity Type of the Provider                                 0
Street Address 1 of the Provider                            0
City of the Provider                                        0
Zip Code of the Provider                                    0
State Code of the Provider                                  0
Country Code of the Provider                                0
Provider Type                                               0
Medicare Participation Indicator                            0
Place of Service                    

## Preprocessing

In [15]:
DropCols = ['index', 'National Provider Identifier',
            'Last Name/Organization Name of the Provider',
            'First Name of the Provider', 'Middle Initial of the Provider',
            'Street Address 1 of the Provider', 'Zip Code of the Provider', 
            "HCPCS Code"]
cleaned_df = cleaned_df.drop(columns=DropCols)



In [16]:
cleaned_df.head()

Credentials of the Provider Gender of the Provider  \
0                        M.D.                      F   
1                        M.D.                      F   
2                         DPM                      M   
3                          MD                      M   
4                          DO                      M   

  Entity Type of the Provider City of the Provider State Code of the Provider  \
0                           I          SAINT LOUIS                         MO   
1                           I         FAYETTEVILLE                         NC   
2                           I          NORTH HAVEN                         CT   
3                           I          KANSAS CITY                         MO   
4                           I              JUPITER                         FL   

  Country Code of the Provider            Provider Type  \
0                           US        Internal Medicine   
1                           US  Obstetrics & Gynecology   
2                           US                 Podiatry   
3                           US        Internal Medicine   
4                           US        Internal Medicine   

  Medicare Participation Indicator Place of Service  \
0                                Y                F   
1                                Y                O   
2                                Y                O   
3                                Y                O   
4                                Y                O   

                                   HCPCS Description HCPCS Drug Indicator  \
0  Initial hospital inpatient care, typically 70 ...                    N   
1  Screening mammography, bilateral (2-view study...                    N   
2  Established patient home visit, typically 25 m...                    N   
3                            Urinalysis, manual test                    N   
4  Injection beneath the skin or into muscle for ...                    N   

  Number of Services Number of Medicare Beneficiaries  \
0                 27                               24   
1                175                              175   
2                 32                               13   
3                 20                               18   
4                 33                               24   

  Number of Distinct Medicare Beneficiary/Per Day Services  \
0                                                 27         
1                                                175         
2                                                 32         
3                                                 20         
4                                                 31         

  Average Medicare Allowed Amount Average Submitted Charge Amount  \
0                    200.58777778                    305.21111111   
1                          123.73                           548.8   
2                           90.65                             155   
3                             3.5                               5   
4                           26.52                              40   

  Average Medicare Payment Amount Average Medicare Standardized Amount  
0                    157.26222222                         160.90888889  
1                          118.83                         135.31525714  
2                      64.4396875                           60.5959375  
3                            3.43                                 3.43  
4                    19.539393939                         19.057575758

In [17]:
# List of columns to convert
columns_to_convert = [
    'Number of Services',
    'Number of Medicare Beneficiaries',
    'Number of Distinct Medicare Beneficiary/Per Day Services',
    'Average Medicare Allowed Amount',
    'Average Submitted Charge Amount',
    'Average Medicare Payment Amount',
    'Average Medicare Standardized Amount'
]

# Convert columns to numeric, setting invalid parsing to NaN
for column in columns_to_convert:
    cleaned_df[column] = pd.to_numeric(cleaned_df[column], errors='coerce')

# Display info to verify the changes
print(cleaned_df[columns_to_convert].info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 7 columns):
 #   Column                                                    Non-Null Count  Dtype  
---  ------                                                    --------------  -----  
 0   Number of Services                                        97347 non-null  float64
 1   Number of Medicare Beneficiaries                          99595 non-null  float64
 2   Number of Distinct Medicare Beneficiary/Per Day Services  98500 non-null  float64
 3   Average Medicare Allowed Amount                           99255 non-null  float64
 4   Average Submitted Charge Amount                           93277 non-null  float64
 5   Average Medicare Payment Amount                           99534 non-null  float64
 6   Average Medicare Standardized Amount                      99530 non-null  float64
dtypes: float64(7)
memory usage: 5.3 MB
None


In [18]:
 cleaned_df

Credentials of the Provider Gender of the Provider  \
0                            M.D.                      F   
1                            M.D.                      F   
2                             DPM                      M   
3                              MD                      M   
4                              DO                      M   
...                           ...                    ...   
99995                          PT                      F   
99996                        ARNP                      F   
99997                        M.D.                      M   
99998                          MD                      F   
99999                          MD                      F   

      Entity Type of the Provider City of the Provider  \
0                               I          SAINT LOUIS   
1                               I         FAYETTEVILLE   
2                               I          NORTH HAVEN   
3                               I          KANSAS CITY   
4                               I              JUPITER   
...                           ...                  ...   
99995                           I           WILMINGTON   
99996                           I              REDMOND   
99997                           I          SAINT LOUIS   
99998                           I            LARCHMONT   
99999                           I           GREAT NECK   

      State Code of the Provider Country Code of the Provider  \
0                             MO                           US   
1                             NC                           US   
2                             CT                           US   
3                             MO                           US   
4                             FL                           US   
...                          ...                          ...   
99995                         IL                           US   
99996                         OR                           US   
99997                         MO                           US   
99998                         NY                           US   
99999                         NY                           US   

                                Provider Type  \
0                           Internal Medicine   
1                     Obstetrics & Gynecology   
2                                    Podiatry   
3                           Internal Medicine   
4                           Internal Medicine   
...                                       ...   
99995  Physical Therapist in Private Practice   
99996                      Nurse Practitioner   
99997                              Cardiology   
99998                       Internal Medicine   
99999  Physical Therapist in Private Practice   

      Medicare Participation Indicator Place of Service  \
0                                    Y                F   
1                                    Y                O   
2                                    Y                O   
3                                    Y                O   
4                                    Y                O   
...                                ...              ...   
99995                                Y                O   
99996                                Y                O   
99997                                Y                F   
99998                                Y                O   
99999                                Y                O   

                                       HCPCS Description HCPCS Drug Indicator  \
0      Initial hospital inpatient care, typically 70 ...                    N   
1      Screening mammography, bilateral (2-view study...                    N   
2      Established patient home visit, typically 25 m...                    N   
3                                Urinalysis, manual test                    N   
4      Injection beneath the skin or into muscle for ...                    N   
...                    

In [19]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import MinMaxScaler,StandardScaler,OneHotEncoder
from sklearn.pipeline import make_pipeline

In [20]:
import pandas as pd

# Assuming 'cleaned_df' is your DataFrame and 'target_column' is your target variable
nominal_columns = ['Credentials of the Provider','Gender of the Provider','Entity Type of the Provider','City of the Provider','State Code of the Provider',
           'Country Code of the Provider','Provider Type','Medicare Participation Indicator','Place of Service','HCPCS Description','HCPCS Drug Indicator']  # Replace with actual nominal columns

# Check unique values in each nominal column
for col in nominal_columns:
    unique_count = cleaned_df[col].nunique()
    print(f"Column '{col}' has {unique_count} unique values.")
    
    if unique_count > 10:
        print(f"Applying frequency encoding to column: {col}")
        # Apply Frequency Encoding
        freq_map = cleaned_df[col].value_counts().to_dict()
        cleaned_df[col] = cleaned_df[col].map(freq_map)
    else:
        print(f"Applying one-hot encoding to column: {col}")
        # Apply One-Hot Encoding for low-cardinality columns
        one_hot = pd.get_dummies(cleaned_df[col], prefix=col)
        cleaned_df = pd.concat([cleaned_df, one_hot], axis=1)
        cleaned_df.drop(columns=[col], inplace=True)

# Display the updated DataFrame
print(cleaned_df.info())


Column 'Credentials of the Provider' has 1854 unique values.
Applying frequency encoding to column: Credentials of the Provider
Column 'Gender of the Provider' has 2 unique values.
Applying one-hot encoding to column: Gender of the Provider
Column 'Entity Type of the Provider' has 2 unique values.
Applying one-hot encoding to column: Entity Type of the Provider
Column 'City of the Provider' has 5846 unique values.
Applying frequency encoding to column: City of the Provider
Column 'State Code of the Provider' has 58 unique values.
Applying frequency encoding to column: State Code of the Provider
Column 'Country Code of the Provider' has 4 unique values.
Applying one-hot encoding to column: Country Code of the Provider
Column 'Provider Type' has 90 unique values.
Applying frequency encoding to column: Provider Type
Column 'Medicare Participation Indicator' has 2 unique values.
Applying one-hot encoding to column: Medicare Participation Indicator
Column 'Place of Service' has 2 unique val

In [21]:
# Convert all boolean columns to integers
boolean_columns = cleaned_df.select_dtypes(include=['bool']).columns
cleaned_df[boolean_columns] = cleaned_df[boolean_columns].astype(float)


In [22]:
cleaned_df

Credentials of the Provider  City of the Provider  \
0                            32757                   500   
1                            32757                   209   
2                             1330                    10   
3                            40083                   317   
4                             2478                    51   
...                            ...                   ...   
99995                          800                   323   
99996                          463                    14   
99997                        32757                   500   
99998                        40083                     6   
99999                        40083                    69   

       State Code of the Provider  Provider Type  HCPCS Description  \
0                            1997          11366               1297   
1                            3725           1028                243   
2                            1403           2027                 44   
3                            1997          11366                460   
4                            7263          11366                732   
...                           ...            ...                ...   
99995                        4073           2780                330   
99996                        1046           5713               4578   
99997                        1997           4602                 75   
99998                        6361          11366               1444   
99999                        6361           2780                319   

       Number of Services  Number of Medicare Beneficiaries  \
0                    27.0                              24.0   
1                   175.0                             175.0   
2                    32.0                              13.0   
3                    20.0                              18.0   
4                    33.0                              24.0   
...                   ...                               ...   
99995                20.0                              20.0   
99996               136.0                             107.0   
99997                11.0                              11.0   
99998                12.0                              12.0   
99999               589.0                              76.0   

       Number of Distinct Medicare Beneficiary/Per Day Services  \
0                                                   27.0          
1                                                  175.0          
2                                                   32.0          
3                                                   20.0          
4                                                   31.0          
...                                                  ...          
99995                                               20.0          
99996                                              136.0          
99997                                               11.0          
99998                                               12.0          
99999                                              587.0          

       Average Medicare Allowed Amount  Average Submitted Charge Amount  ...  \
0                           200.587778                       305.211111  ...   
1                           123.730000                       548.800000  ...   
2                            90.650000                       155.000000  ...   
3                             3.500000                         5.000000  ...   
4                            26.520000                        40.000000  ...   
...                                ...                              ...  ...   
99995                        85.372500                       214.000000  ...   
99996                        61.270000                       144.051471  ...   
99997                        17.980000                       109.545455  ...   
99998                        30.540000                        65.000000

In [23]:
cleaned_df.fillna(cleaned_df.mean(), inplace=True)

In [24]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 26 columns):
 #   Column                                                    Non-Null Count   Dtype  
---  ------                                                    --------------   -----  
 0   Credentials of the Provider                               100000 non-null  int64  
 1   City of the Provider                                      100000 non-null  int64  
 2   State Code of the Provider                                100000 non-null  int64  
 3   Provider Type                                             100000 non-null  int64  
 4   HCPCS Description                                         100000 non-null  int64  
 5   Number of Services                                        100000 non-null  float64
 6   Number of Medicare Beneficiaries                          100000 non-null  float64
 7   Number of Distinct Medicare Beneficiary/Per Day Services  100000 non-null  float64
 8   Avera

In [25]:
min_max_f=['Number of Services','Number of Medicare Beneficiaries','Number of Distinct Medicare Beneficiary/Per Day Services']
Standard_f=['Average Medicare Allowed Amount','Average Submitted Charge Amount','Average Medicare Payment Amount','Average Medicare Standardized Amount']


In [26]:
min_max_transform=make_pipeline(MinMaxScaler())
Standard_transform=make_pipeline(StandardScaler())

In [27]:
preprocessor=ColumnTransformer(transformers=[
    ('min max',min_max_transform,min_max_f),
    ('standard',Standard_transform,Standard_f)
],remainder='passthrough',verbose_feature_names_out=False,force_int_remainder_cols=False)

In [28]:
preprocessor.set_output(transform='pandas')

ColumnTransformer(force_int_remainder_cols=False, remainder='passthrough',
                  transformers=[('min max',
                                 Pipeline(steps=[('minmaxscaler',
                                                  MinMaxScaler())]),
                                 ['Number of Services',
                                  'Number of Medicare Beneficiaries',
                                  'Number of Distinct Medicare Beneficiary/Per '
                                  'Day Services']),
                                ('standard',
                                 Pipeline(steps=[('standardscaler',
                                                  StandardScaler())]),
                                 ['Average Medicare Allowed Amount',
                                  'Average Submitted Charge Amount',
                                  'Average Medicare Payment Amount',
                                  'Average Medicare Standardized Amount'])],
                  verbose_feature_names_out=False)

In [29]:
transformed_df=preprocessor.fit_transform(cleaned_df)

In [30]:
transformed_df

Number of Services  Number of Medicare Beneficiaries  \
0                0.016194                          0.013158   
1                0.165992                          0.165992   
2                0.021255                          0.002024   
3                0.009109                          0.007085   
4                0.022267                          0.013158   
...                   ...                               ...   
99995            0.009109                          0.009109   
99996            0.126518                          0.097166   
99997            0.000000                          0.000000   
99998            0.001012                          0.001012   
99999            0.585020                          0.065789   

       Number of Distinct Medicare Beneficiary/Per Day Services  \
0                                               0.016194          
1                                               0.165992          
2                                               0.021255          
3                                               0.009109          
4                                               0.020243          
...                                                  ...          
99995                                           0.009109          
99996                                           0.126518          
99997                                           0.000000          
99998                                           0.001012          
99999                                           0.582996          

       Average Medicare Allowed Amount  Average Submitted Charge Amount  \
0                             1.098226                         0.621012   
1                             0.352134                         1.940981   
2                             0.031012                        -0.192958   
3                            -0.814992                        -1.005784   
4                            -0.591527                        -0.816125   
...                                ...                              ...   
99995                        -0.020219                         0.126753   
99996                        -0.254193                        -0.252286   
99997                        -0.674428                        -0.439269   
99998                        -0.552503                        -0.680654   
99999                        -0.474250                        -0.778910   

       Average Medicare Payment Amount  Average Medicare Standardized Amount  \
0                             0.972452                              1.003321   
1                             0.549955                              0.722789   
2                            -0.047975                             -0.096209   
3                            -0.718674                             -0.722804   
4                            -0.541578                             -0.551510   
...                                ...                                   ...   
99995                        -0.088807                             -0.078095   
99996                        -0.426514                             -0.354403   
99997                        -0.601485                             -0.600151   
99998                        -0.427351                             -0.482868   
99999                        -0.429474                             -0.476378   

       Credentials of the Provider  City of the Provider  \
0                            32757                   500   
1                            32757                   209   
2                             1330                    10   
3                            40083                   317   
4                             2478                    51   
...                            ...                   ...   
99995                          800                   323   
99996                          463                    14   
99997                        32757         